# Exploración inicial de datos
Por Mateo Soto

Última actualización: 19/Ago/2025

## Description
Data overview and exploration to check data types and fix any issue with the data,
prior to a correct data analysis and visualization of the Boston housing price dataset.


## 📚 Import  libraries

In [ ]:
# base libraries for data science
from pathlib import Path

import pandas as pd
import pyarrow as pa

## 💾 Load data

In [2]:
# data directory path
DATA_DIR = Path.cwd().resolve().parents[1] / "data"

boston_df = pd.read_csv(DATA_DIR / "01_raw/Precios_Casas_Boston.csv")

In [3]:
# data description
boston_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 448 entries, 0 to 447
Data columns (total 15 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   ID       448 non-null    int64  
 1   crim     443 non-null    float64
 2   zn       446 non-null    float64
 3   indus    442 non-null    float64
 4   chas     446 non-null    float64
 5   nox      444 non-null    float64
 6   rm       442 non-null    float64
 7   age      444 non-null    float64
 8   dis      442 non-null    float64
 9   rad      444 non-null    float64
 10  tax      443 non-null    float64
 11  ptratio  446 non-null    float64
 12  black    445 non-null    float64
 13  lstat    447 non-null    float64
 14  medv     432 non-null    float64
dtypes: float64(14), int64(1)
memory usage: 52.6 KB


In [4]:
boston_df.sample(10)

,ID,crim,zn,indus,chas,nox,rm,age,dis,rad,tax,ptratio,black,lstat,medv
353,7,0.08829,12.5,7.87,0.0,524.00,6012.0,66.6,5.5605,5.0,311.0,15.2,395.60,12.43,22.9
140,202,0.03445,82.5,2.03,0.0,415.00,6162.0,38.4,6.2700,2.0,348.0,14.7,393.77,7.43,24.1
438,176,0.06664,0.0,4.05,0.0,0.51,6546.0,33.1,3.1323,5.0,296.0,16.6,390.96,5.33,29.4
428,165,2.24236,0.0,19.58,0.0,605.00,5854.0,91.8,2422.0000,5.0,403.0,14.7,395.11,11.64,22.7
439,485,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,20.2,370.73,13.34,NaN
366,303,0.09266,34.0,6.09,0.0,433.00,6495.0,18.4,5.4917,7.0,329.0,16.1,383.61,8.67,26.4
272,420,11.81230,0.0,18.10,0.0,718.00,6824.0,76.5,1794.0000,24.0,666.0,20.2,48.45,22.74,8.4
111,165,2.24236,0.0,19.58,0.0,605.00,5854.0,91.8,2422.0000,5.0,403.0,14.7,395.11,11.64,22.7
395,434,5.58107,0.0,18.10,0.0,713.00,6436.0,87.9,2.3158,24.0,666.0,20.2,100.19,16.22,NaN
378,402,14.23620,0.0,18.10,0.0,693.00,6343.0,100.0,1.5741,24.0,666.0,20.2,396.90,20.32,7.2


In [5]:
boston_df.describe()

,ID,crim,zn,indus,chas,nox,rm,age,dis,rad,tax,ptratio,black,lstat,medv
count,448.000000,443.000000,446.000000,442.000000,446.000000,444.000000,442.000000,444.000000,442.000000,444.000000,443.000000,446.000000,445.000000,447.000000,432.000000
mean,263.933036,75.806953,9.149103,12.032986,0.056054,479.452313,5590.948077,71.260135,352.520877,9.801802,418.356659,18.400897,347.803213,13.190761,21.987963
std,150.877878,894.067182,20.794410,7.074594,0.230284,227.804740,1991.240670,27.671128,1165.055292,8.798543,171.611273,2.242069,99.918129,7.243423,9.219574
min,1.000000,0.006320,0.000000,0.740000,0.000000,0.400000,4.970000,6.000000,1.129600,1.000000,188.000000,12.600000,3.500000,1.730000,5.000000
25%,140.000000,0.093780,0.000000,5.860000,0.000000,433.000000,5710.000000,50.675000,2.016200,4.000000,284.500000,16.800000,368.740000,7.525000,16.100000
50%,265.500000,0.352330,0.000000,10.010000,0.000000,532.000000,6098.500000,82.150000,3.099250,5.000000,384.000000,19.000000,391.230000,12.030000,20.600000
75%,410.250000,4.548895,0.000000,18.100000,0.000000,614.000000,6437.750000,95.000000,5.499950,24.000000,666.000000,20.200000,395.990000,17.280000,24.700000
max,516.000000,15288.000000,100.000000,27.740000,1.000000,871.000000,8725.000000,100.000000,8344.000000,24.000000,711.000000,21.200000,396.900000,37.970000,50.000000


In [6]:
boston_df.isna().sum()

ID          0
crim        5
zn          2
indus       6
chas        2
nox         4
rm          6
age         4
dis         6
rad         4
tax         5
ptratio     2
black       3
lstat       1
medv       16
dtype: int64

## Hallazgos iniciales

- El dataset tiene 448 filas y 15 columnas (incluyendo `ID`).
- Todas las columnas se cargaron como `float64` o `int64`; no hay columnas tipo `object` o string.
- Las columnas `crim`, `zn`, `indus`, `chas`, `nox`, `rm`, `age`, `dis`, `rad`, `tax`,
  `medv` presentan valores nulos (entre 1 y 16 registros por columna), representados
  directamente como `NaN` — no hay un valor centinela como `"?"`.
- **Problema de escala detectado**: las columnas `nox` (proporción, rango esperado 0.4–0.9), `rm` (número de cuartos, rango esperado ~3–9), `crim` (proporción de rango esperado ~0.006 – 90 (tasa de criminalidad per cápita))  y `dis` (proporción de rango esperado 	1.13 – 12.13)   tienen valores muy por fuera de rango (Hasta 15,288), consistente con un corrimiento del punto decimal en una parte de los registros. Se corrige en la sección de conversión de tipos.
- `chas` tiene solo 2 valores únicos (0/1) → es en realidad una variable booleana/categórica,
  no continua.
- `rad` tiene 9 valores únicos, incluyendo `24` como posible categoría atípica → revisar
  si es un índice de accesibilidad discreto (categórico ordinal) en vez de numérico continuo.

## Diccionario de variables

| Variable | Nombre completo / Fenómeno que representa | Unidad / Naturaleza |
|---|---|---|
| **ID** | Identificador único del registro (no es un fenómeno real, solo un índice) | — |
| **crim** | Tasa de criminalidad per cápita por ciudad/zona | Delitos per cápita |
| **zn** | Proporción de terreno residencial destinado a lotes grandes (>25,000 pies²) | % del área |
| **indus** | Proporción de acres dedicados a negocios no minoristas (industria) por ciudad | % del área |
| **chas** | Si el terreno limita con el río Charles | Dummy: 1 = sí, 0 = no |
| **nox** | Concentración de óxidos de nitrógeno (contaminación del aire) | Partes por 10 millones (proporción 0–1) |
| **rm** | Número promedio de habitaciones/cuartos por vivienda | Cantidad de cuartos |
| **age** | Proporción de viviendas ocupadas por sus dueños, construidas antes de 1940 | % de viviendas antiguas |
| **dis** | Distancia ponderada a cinco centros de empleo de Boston | Distancia ponderada |
| **rad** | Índice de accesibilidad a autopistas radiales | Índice (categórico ordinal) |
| **tax** | Tasa de impuesto predial de valor completo | USD por cada $10,000 |
| **ptratio** | Ratio alumno-profesor por ciudad/zona escolar | Estudiantes por docente |
| **black** | Medida derivada de la proporción de población afrodescendiente por ciudad: 1000(Bk - 0.63)^2, donde Bk es esa proporción | Índice transformado |
| **lstat** | Porcentaje de población de estatus socioeconómico bajo (*lower status*) | % de la población |
| **medv** | Valor mediano de las viviendas ocupadas por sus dueños — **es la variable objetivo (target)** | Miles de USD |

### Nota histórica sobre la variable `black`

Esta variable es controversial y ampliamente criticada en la literatura de ciencia de datos
moderna. Fue construida por Harrison & Rubinfeld (1978) bajo el supuesto de que la relación
entre la proporción de población afrodescendiente y el valor de la vivienda no es lineal sino
cuadrática, centrada artificialmente en 0.63. Por esta razón, scikit-learn eliminó este dataset
de sus datasets de ejemplo en 2020, citando el sesgo racial codificado directamente en la
construcción de esta variable.

### Supuesto de corrección de escala

Se identificó que un subconjunto de valores en `crim`, `nox`, `rm` y `dis` está fuera del
rango físicamente esperado para cada variable (según las definiciones oficiales del dataset
en `Informacion.txt` y la documentación original de Harrison & Rubinfeld, 1978), de forma
consistente con un corrimiento del punto decimal (× 1000).

**Este comportamiento no está documentado en `Informacion.txt` provisto por el profesor** —
se trata de un hallazgo propio durante la exploración, no de un problema conocido/esperado
del ejercicio. Se asume que estos valores deben dividirse entre 1000 para volver al rango
esperado, dado que las demás columnas (`zn`, `indus`, `age`, `tax`, `ptratio`, `black`,
`lstat`, `medv`) sí caen dentro de sus rangos oficiales sin corrección, lo que sugiere un
error de captura/transformación aislado a estas cuatro columnas específicas y no un problema
generalizado del dataset.

In [7]:
# Comparación de rangos: valores "sanos" vs "atípicos" por columna, para justificar el factor de escala
for col, umbral in [("crim", 100), ("nox", 1), ("rm", 15), ("dis", 20)]:
    sanos = boston_df.loc[boston_df[col] <= umbral, col]
    atipicos = boston_df.loc[boston_df[col] > umbral, col]
    print(
        f"{col}: sanos [{sanos.min():.3f} - {sanos.max():.3f}]  |  atípicos÷1000 [{atipicos.min() / 1000:.3f} - {atipicos.max() / 1000:.3f}]"
    )

crim: sanos [0.006 - 73.534]  |  atípicos÷1000 [0.537 - 15.288]
nox: sanos [0.400 - 0.770]  |  atípicos÷1000 [0.385 - 0.871]
rm: sanos [4.970 - 8.040]  |  atípicos÷1000 [3.561 - 8.725]
dis: sanos [1.130 - 10.710]  |  atípicos÷1000 [1.137 - 8.344]


In [14]:
# Umbrales de escala física esperada por variable (ver hallazgo documentado arriba)
UMBRAL_CRIM = 100  # tasa de criminalidad per cápita
UMBRAL_RM = 15  # cuartos promedio por vivienda
UMBRAL_DIS = 20  # distancia ponderada a centros de empleo
FACTOR_CORRECCION = 1000

boston_df.loc[boston_df["crim"] > UMBRAL_CRIM, "crim"] = boston_df["crim"] / FACTOR_CORRECCION
boston_df.loc[boston_df["rm"] > UMBRAL_RM, "rm"] = boston_df["rm"] / FACTOR_CORRECCION
boston_df.loc[boston_df["dis"] > UMBRAL_DIS, "dis"] = boston_df["dis"] / FACTOR_CORRECCION

boston_df[["crim", "nox", "rm", "dis"]].describe()

,crim,nox,rm,dis
count,443.000000,444.000000,442.000000,442.000000
mean,3.856402,0.571313,6.232952,3.492427
std,8.211991,0.117252,0.683374,1.895131
min,0.006320,0.385000,3.561000,1.129600
25%,0.093780,0.472000,5.875000,1.980425
50%,0.352330,0.544000,6.152000,2.781150
75%,4.149860,0.647000,6.538000,4.560125
max,73.534100,0.871000,8.725000,10.710300


## Categorización de variables

### Nominal / Booleana
- `chas`: Limita con el río Charles (dummy variable)
  - 0 = No
  - 1 = Sí

### Ordinal
- `rad`: Índice de accesibilidad a autopistas radiales (categórico ordinal, valores discretos)

### Numéricas continuas
crim, zn, indus, nox, rm, age, dis, tax, ptratio, black, lstat, medv

### Identificador (no predictivo)
- `ID`

In [9]:
cols_boolean = ["chas"]
boston_df[cols_boolean] = boston_df[cols_boolean].astype(
    "boolean"
)  # nullable boolean, preserva NaN

In [10]:
cols_categoric_ordinal = ["rad"]
boston_df[cols_categoric_ordinal] = boston_df[cols_categoric_ordinal].astype(
    "Int64"
)  # nullable int, preserva NaN

In [11]:
cols_numeric_float = [
    "crim",
    "zn",
    "indus",
    "nox",
    "rm",
    "age",
    "dis",
    "tax",
    "ptratio",
    "black",
    "lstat",
    "medv",
]
boston_df[cols_numeric_float] = boston_df[cols_numeric_float].astype("float64")

In [12]:
boston_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 448 entries, 0 to 447
Data columns (total 15 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   ID       448 non-null    int64  
 1   crim     443 non-null    float64
 2   zn       446 non-null    float64
 3   indus    442 non-null    float64
 4   chas     446 non-null    boolean
 5   nox      444 non-null    float64
 6   rm       442 non-null    float64
 7   age      444 non-null    float64
 8   dis      442 non-null    float64
 9   rad      444 non-null    Int64  
 10  tax      443 non-null    float64
 11  ptratio  446 non-null    float64
 12  black    445 non-null    float64
 13  lstat    447 non-null    float64
 14  medv     432 non-null    float64
dtypes: Int64(1), boolean(1), float64(12), int64(1)
memory usage: 50.4 KB


In [13]:
schema = pa.Table.from_pandas(boston_df).schema

boston_df.to_parquet(
    DATA_DIR / "02_intermediate/boston_type_fixed.parquet", index=False, schema=schema
)

## 📊 Analysis of Results and Conclusions 

Description of the results obtained and if there are conclusions that can be drawn from them.

The analysis of results must be related to the description of the task.

**Note:** An analysis of results does not necessarily lead to conclusions, but to ideas or proposals for future work


Se corrigió la escala inconsistente en `crim`, `nox`, `rm` y `dis`, luego de haer un analisis y verificar la probabilidad de un error de "," en los valores, dividiendolo por 1000 y ubicando los errores en el margen plausible dentro de las caractersiticas de los datos. 
 se tipificó `chas` como booleana
y `rad` como entero categórico ordinal, y se preservaron los valores nulos usando tipos
"nullable" de pandas (`boolean`, `Int64`) en vez de eliminarlos, dado que en este punto
del proyecto aún no se ha definido la estrategia de imputación (se abordará en Feature
Engineering). El dataset queda guardado con su schema de pyarrow explícito para
análisis posterior.